### PR TITLE
Authentication

### LINKED ISSUE #35: Support X.509 credentials

### PR DESCRIPTION
**Pull request by bifurcation**
~**DO NOT MERGE: NEEDS TESTS**~ (thanks @suhasHere)

This PR provides the tools for applications to add authentication to MLS.  

* States track which leaves got new credentials in the transition to this state
* X.509 certificate chains can be used as credentials

Note that this approach does **not** have the MLS stack perform authentication.  It's not even as integrate as the typical approach in TLS stacks, where the application provides a certificate callback.  Instead, the presumption is that the application will observe which leaves in the tree have new (and thus unverified) credentials via the API provided, then validate those credentials at its leisure. 

Fixes #35

### CODE REVIEW

File: credential.go [hunk 1]
```diff
@@ -28,8 +34,165 @@ type BasicCredential struct {
...
+	pool := newCertPool(trusted)
+
+	var curr, next *x509.Certificate
+	for i := 0; i < len(cred.Chain)-1; i++ {
+		curr = cred.Chain[i]
+		next = cred.Chain[i+1]
+
+		// If there is a valid signature from a trusted certificate, the chain is valid
+		parent, ok := pool.parent(curr)
+		if ok && cred.Chain[i].CheckSignatureFrom(parent) == nil {
```
Comment (suhasHere) [inline credential.go:176]: can be curr instead of cred.Chain[i]

File: state.go [hunk 2]
```diff
@@ -478,6 +481,15 @@ func (s *State) applyUpdateProposal(target LeafIndex, update *UpdateProposal) er
...
 
+	currKP, ok := s.Tree.KeyPackage(target)
+	if !ok {
+		return fmt.Errorf("mls.state: Attempt to update an empty leaf")
+	}
+
+	if !update.KeyPackage.Credential.Equals(currKP.Credential) {
+		s.NewCredentials[target] = true
+	}
+
```
Comment (suhasHere) [inline state.go:492]: i wonder if we update the bool map after we perform s.Tree.UpdateLeaf ? aren't we checking the updated credentil before doing the update the way it is coded now ?

File: credential.go (hunk 1 above)
Comment (bifurcation) [reply credential.go:176]: Good point, thanks.

File: state.go (hunk 2 above)
Comment (bifurcation) [reply state.go:492]: We are not checking the credential before doing the update.  Credential checking here is lazy, after all the changes have been done.

Comment (bifurcation) [reply state.go:492]: Whether that's a good idea, we should discuss later :)

**Pull request merged**